src/box/txn_limbo.c

```diff
@@ -1304,6 +1338,134 @@ txn_limbo_update_system_spaces_is_sync_state(struct txn_limbo *limbo,
 	system_spaces_update_is_sync_state(is_promote);
 }
 
+static int
+txn_limbo_entry_iterator_next(struct port_c_iterator *it, struct port *out,
+			      bool *is_eof)
+{
+	struct txn_limbo_entry_port_c_iterator *entry_it =
+		(struct txn_limbo_entry_port_c_iterator *)it;
+	struct txn_stmt *stmt = entry_it->stmt;
+	if (stmt == NULL) {
+		*is_eof = true;
+		return 0;
+	}
+
+	*is_eof = false;
+	struct xrow_header *row = stmt->row;
+	port_c_create(out);
+	/*
+	 * The iterator returns the following transaction statement information:
+	 *  1. An ordinal request number;
+	 *  2. The term during which the transaction occurred.
+	 *  3. The ID of the transaction;
+	 *  4. The ID of the limbo owner or 0 for local rows that are part of a
+	 *  global transaction;
+	 *  5. The LSN of the request.
+	 *  6. The ID of the space to which the request was applied.
+	 *  7. The request type.
+	 *  8. The argument of the request depending on the operation type
+	 *  9. An optional argument for the UPDATE/UPSERT request, may be nil.
+	 *  10. The tuple which was replaced by the operation, may be nil.
+	 *  11. The tuple created by the operation, may be nil.
```

Comment:
This design mirrors the design of `on_commit`/`on_rollback` iterators [^1]:
> The iterator has:
> 
> * an ordinal request number
> * the old value of the tuple before the request (nil for an insert request)
> * the new value of the tuple after the request (nil for a delete request)
> * the ID of the space

[^1]: https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_txn_management/on_commit/

Reply:
/cc @sergepetrenko @Serpentian

Reply:
Yeah, but in case of on_commit/rollback it is 4 values, while in the new API it is 11 values. This is quite a big difference. We can't expose this kind of multireturn, it is too complex. Even in case of on_commit/rollback I would now say it should have been an object, because now we can't add stuff to it (I know that I did those triggers, but I didn't know any better back then 😭).

Reply:
<img width="946" alt="Снимок экрана 2024-12-17 в 12 34 27" src="https://github.com/user-attachments/assets/16d80b0a-dfc1-45b7-94cc-131c325b3e43" />

Just for the record, here's the format of repair_queue space we've agreed upon. Only `term` is missing, and it's needed, obviously.

From what I can see, `id` and `transaction_id` may be calculated by the trigger itself, no need to pass them,
replica_id, lsn, space_id, operation_type, key/tuple, ops_array, old_tuple, new_tuple may all be passed in a format similar to xlog iterator. Here's the output of `xlog.pairs` when `wal_ext` is enabled:
```
tarantool> for k,v in xlog.pairs('00000000000000000000.xlog') do json = require'json' print(json.encode(k)) print(json.encode(v)) end
1
{"HEADER":{"lsn":1,"replica_id":1,"type":"INSERT","timestamp":1734430078.2496},"BODY":{"space_id":280,"new_tuple":[512,1,"test","memtx",0,{},[]],"tuple":[512,1,"test","memtx",0,{},[]]}}
2
{"HEADER":{"lsn":2,"replica_id":1,"type":"INSERT","timestamp":1734430082.6448},"BODY":{"space_id":288,"new_tuple":[512,0,"pk","tree",{"unique":true},[[0,"unsigned"]]],"tuple":[512,0,"pk","tree",{"unique":true},[[0,"unsigned"]]]}}
3
{"HEADER":{"lsn":3,"replica_id":1,"type":"INSERT","timestamp":1734430089.2591},"BODY":{"space_id":512,"new_tuple":[1,"asdfb"],"tuple":[1,"asdfb"]}}
4
{"HEADER":{"lsn":4,"replica_id":1,"type":"INSERT","timestamp":1734430094.6888},"BODY":{"space_id":512,"new_tuple":[2,"bsdfa"],"tuple":[2,"bsdfa"]}}
5
{"HEADER":{"lsn":5,"replica_id":1,"type":"REPLACE","timestamp":1734430098.0942},"BODY":{"space_id":512,"old_tuple":[2,"bsdfa"],"new_tuple":[2],"tuple":[2]}}
6
{"HEADER":{"lsn":6,"replica_id":1,"type":"UPDATE","timestamp":1734430107.2436},"BODY":{"new_tuple":[1,"asdfb","bababa"],"old_tuple":[1,"asdfb"],"space_id":512,"index_base":1,"key":[1],"tuple":[["=",3,"bababa"]]}}
---
...
```

So I propose the following format:
```yaml
obj = {
    HEADER = {
        lsn = ...,
        replica_id = ...,
        type = ...,
        term = ...
    },
    BODY = {
        space_id = ...,
        key = ...,
        tuple = ...,
        old_tuple = ...,
        new_tuple = ...,
    }
}
```

Reply:
I am super super sorry 🙏, but how about we also change the repair queue format to store rows as objects? And only put into the space fields the things like `id`? The indexes anyway can be by json paths, if we need to index here anything. Otherwise I am really scared for extendibility of the format.

Speaking of the iteration: do we really want to make it so binary like `HEADER` and `BODY`? We are not iterating over a file here. And it doesn't reflect the binary structure anyway, because we don't store `term` in xrows, but above it looks like stored in the header.

But I like the rows being objects much more than multireturns 🔥! A couple of suggestions based on the things I wrote above:
- Perhaps flatten the object? So the header and body keys are all in one level? The motivation is what I said about the iterator being over txn statements, not xlogs.
- If not flatten, then perhaps add the term at the same level as `HEADER` and `BODY` keys? Otherwise it isn't what xlog iterator would return.

Thanks for the feedback and the discussion, @sergepetrenko and @CuriousGeorgiy , and apologies once again that I didn't realize those things during the RFC review. I hope not much would need change in the code.

Reply:
> I am super super sorry 🙏, but how about we also change the repair queue format to store rows as objects? And only put into the space fields the things like id? The indexes anyway can be by json paths, if we need to index here anything. Otherwise I am really scared for extendibility of the format.

I agree, this makes sense not to introduce huge space formats with lots of fields. We even have this ticket regarding a similar problem in space `_func`: https://github.com/tarantool/tarantool/issues/9894

I suggest the following format then:
`[id : unsigned, transaction_id : unsigned, row_meta : map, statement : map]`
`row_meta` would contain `replica_id`, `lsn`, `term`.
`statement` would contain the rest: `op_type`, `space_id`, `key`, `tuple`, `old_tuple`, `new_tuple`.

> Perhaps flatten the object? So the header and body keys are all in one level? The motivation is what I said about the iterator being over txn statements, not xlogs.

I'd rather not mix row meta and operation details. It looks strange to pile everything in one level IMO.

> If not flatten, then perhaps add the term at the same level as HEADER and BODY keys? Otherwise it isn't what xlog iterator would return.

Maybe make the  object format similar to space format I propose, then? In this case it would return objects almost in the same format in which they will be stored in the space (`id` and `transaction_id` will be missing, since we agreed the user could calculate them himself). Does it make sense?
```
obj = {
    row_meta = {
        replica_id = ...,
        lsn = ...,
        term = ...,
    },
    statement = {
        op_type = ...,
        space_id = ...,
        key = ...,
        tuple = ...,
        old_tuple = ...,
        new_tuple = ...,
    }
}
```

Reply:
Thanks for considering it! Yes, looks way better. I would even suggest to shorten `row_meta` to just `meta`, otherwise top 💪.